src/libs/SubscriptionUtils.ts

```diff
@@ -30,25 +119,231 @@ Onyx.connect({
+ */
+function hasAmountOwed(): boolean {
+    return !!amountOwed;
+}
+
+/**
+ * @returns Whether there is a card authentication error.
+ */
+function hasCardAuthenticatedError() {
+    return stripeCustomerId?.status === 'authentication_required' && amountOwed === 0;
+}
+
+/**
+ * @returns Whether there is a billing dispute pending.
+ */
+function hasBillingDisputePending() {
+    return !!billingDisputePending ?? false;
+}
+
+/**
+ * @returns Whether there is a card expired error.
+ */
+function hasCardExpiredError() {
+    return billingStatus?.declineReason === 'expired_card' && amountOwed !== 0;
+}
+
+/**
+ * @returns Whether there is an insufficient funds error.
+ */
+function hasInsufficientFundsError() {
+    return billingStatus?.declineReason === 'insufficient_funds' && amountOwed !== 0;
+}
+
+/**
+ * @returns The card to be used for subscription billing.
+ */
+function getCardForSubscriptionBilling(): Fund | undefined {
+    return Object.values(fundList ?? {}).find((card) => card?.isDefault);
+}
+
+/**
+ * @returns Whether the card is due to expire soon.
+ */
+function hasCardExpiringSoon(): boolean {
+    if (!isEmptyObject(billingStatus)) {
+        return false;
+    }
+
+    const card = getCardForSubscriptionBilling();
+
+    if (!card) {
+        return false;
+    }
+
+    const cardYear = card?.accountData?.cardYear;
+    const cardMonth = card?.accountData?.cardMonth;
+    const currentYear = new Date().getFullYear();
+    const currentMonth = new Date().getMonth();
+
+    const isExpiringThisMonth = cardYear === currentYear && cardMonth === currentMonth;
+    const isExpiringNextMonth = cardYear === (currentMonth === 12 ? currentYear + 1 : currentYear) && cardMonth === (currentMonth === 12 ? 1 : currentMonth + 1);
+
+    return isExpiringThisMonth || isExpiringNextMonth;
+}
+
+/**
+ * @returns Whether there is a retry billing error.
+ */
+function hasRetryBillingError(): boolean {
+    return !!retryBillingFailed ?? false;
+}
+
+/**
+ * @returns Whether the retry billing was successful.
+ */
+function isRetryBillingSuccessful(): boolean {
+    return !!retryBillingSuccessful ?? false;
+}
+
+type SubscriptionStatus = {
+    status: string;
+    isError?: boolean;
+};
+
+/**
+ * @returns The subscription status.
+ */
+function getSubscriptionStatus(): SubscriptionStatus | undefined {
+    if (hasOverdueGracePeriod()) {
+        if (hasAmountOwed()) {
+            // 1. Policy owner with amount owed, within grace period
+            if (!hasGracePeriodOverdue()) {
+                return {
+                    status: PAYMENT_STATUS.POLICY_OWNER_WITH_AMOUNT_OWED,
+                    isError: true,
+                };
+            }
+
+            // 2. Policy owner with amount owed, overdue (past grace period)
+            if (hasGracePeriodOverdue()) {
+                return {
+                    status: PAYMENT_STATUS.POLICY_OWNER_WITH_AMOUNT_OWED_OVERDUE,
```

Comment:
We should also add `isError`: true here.
More details: https://github.com/Expensify/App/issues/52598#issuecomment-2477342129